core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java

```diff
@@ -309,11 +312,22 @@ private static Object transformTemplateParamValue(Object value) {
         return value;
     }
 
-    private static Optional<Object> memoryId(AiServiceMethodCreateInfo createInfo, Object[] methodArgs) {
+    private static Object memoryId(AiServiceMethodCreateInfo createInfo, Object[] methodArgs, boolean hasChatMemoryProvider) {
         if (createInfo.getMemoryIdParamPosition().isPresent()) {
-            return Optional.of(methodArgs[createInfo.getMemoryIdParamPosition().get()]);
+            return methodArgs[createInfo.getMemoryIdParamPosition().get()];
         }
-        return Optional.empty();
+        if (hasChatMemoryProvider) {
+            // first we try to use the current context in order to make sure that we don't interleave chat messages of concurrent requests
+            ArcContainer container = Arc.container();
+            if (container != null) {
+                ManagedContext requestContext = container.requestContext();
+                if (requestContext.isActive()) {
+                    return requestContext;
```

Comment:
Stupid question: what's your contract for the memory id? Does it use equals/hashCode? If so does the requestContext implement it properly?

Do you store the memoryId somewhere? Would it be a problem to keep the request context around if so?

Probably questions you already answered but I couldn't help thinking about them when reading the patch.

Reply:
> what's your contract for the memory id?

It can be any object

 > If so does the requestContext implement it properly?

It does not implement them at all

> Do you store the memoryId somewhere? Would it be a problem to keep the request context around if so?

Those exact questions plagued us for a while, but they have since been sorted out with the proper integration between the scope of the AiService and the lifecycle of the request.

> Probably questions you already answered but I couldn't help thinking about them when reading the patch.

👍🏼